ansible_rulebook/engine.py

```diff
@@ -293,32 +298,29 @@ async def run_actions(self, data: Dict):
             finally:
                 logger.debug(lang.get_pending_events(self.name))
 
-            while not self.ruleset_queue_plan.plan.queue.empty():
-                matched = True
-                item = cast(
-                    ActionContext,
-                    await self.ruleset_queue_plan.plan.queue.get(),
-                )
-                result = await self.call_action(*item)
-                results.append(result)
-
-            if matched:
-                # TODO: how useful to log the results without logging the
-                # original event data?
-                await self.event_log.put(
-                    dict(type="ProcessedEvent", results=results)
-                )
-        except MessageNotHandledException:
-            logger.info("MessageNotHandledException: %s", data)
-            # TODO: is it really necessary to add such event to event_log?
-            await self.event_log.put(dict(type="MessageNotHandled"))
-        except ShutdownException:
-            await self.ruleset_queue_plan.source_queue.put(Shutdown())
-        except Exception:
-            logger.exception("Error processing %s", data)
-
-    async def call_action(
+    async def _drain_actionplan_queue(self):
+        logger.info("Waiting for actions on events from %s", self.name)
+        action_id = 0
+        while not self.stopped:
+            queue_item = await self.ruleset_queue_plan.plan.queue.get()
+            if isinstance(queue_item, Shutdown):
```

Comment:
@bzwei In the action plan queue who inserts a ShutDown Message? I thought they would always be ActionContext coming from here.
https://github.com/ansible/ansible-rulebook/blob/bfe9f6ebdbc42e39aa793eeb0311af00e2425c09/ansible_rulebook/rule_generator.py#L63

Reply:
No. Shutdown is a special indication of ending the source https://github.com/ansible/ansible-rulebook/blob/main/ansible_rulebook/engine.py#L174. There is no ActionContext for shutdown.

Reply:
@bzwei if the Plan.Queue is filled with other ActionItems you would have to drain all of them before you get to the Shutdown. The same happens even in the Playbook serialization Queue. If there are multiple entries in the Playbook queue we would not be able to get to the Shutdown or Stop till all the items have been drained.